### PR TITLE
FVM integration

### PR DESCRIPTION
**Pull request by raulk**

### CODE REVIEW

File: rust/src/fvm/machine.rs [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    let next_id = NEXT_ID.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+    let mut machines = FVM_MAP.lock().unwrap();
+    machines.insert(next_id, machine);
+    next_id
+}
+
+fn get_default_config() -> fvm::Config {
+    Config {
+        initial_pages: 1024, //FIXME
+        max_pages: 32768,    // FIXME
```
Comment (raulk) [inline rust/src/fvm/machine.rs:50]: @Stebalien any sense of the values we want to set here? Noting that WASM page size is 64KiB, so we may want to start with 16 pages (1MiB), and have no upper limit for M0? We probaly want to instrument the WASM runtime to record the pages we ended up allocating with every execution.

File: rust/src/fvm/machine.rs [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+#[cfg(not(target_os = "windows"))]
+pub unsafe extern "C" fn fil_create_fvm_machine(
+    fvm_version: fil_FvmRegisteredVersion,
+    chain_epoch: u64,
+    token_amount: u64,
+    network_version: u64,
+    state_root_ptr: *const u8,
+    state_root_len: libc::size_t,
+    blockstore_id: u64,
+    externs_id: u64,
```
Comment (raulk) [inline rust/src/fvm/machine.rs:69]: I _think_ externs can be stateless if the blockstore is _not_ folded into the externs, but need to think more about this.

File: rust/src/fvm/machine.rs [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+/// Note: the incoming args as u64 and odd conversions to i32/i64
+/// for some types is due to the generated bindings not liking the
+/// 32bit types as incoming args
+///
+#[no_mangle]
+#[cfg(not(target_os = "windows"))]
+pub unsafe extern "C" fn fil_create_fvm_machine(
+    fvm_version: fil_FvmRegisteredVersion,
+    chain_epoch: u64,
+    token_amount: u64,
```
Comment (raulk) [inline rust/src/fvm/machine.rs:64]: Is this the basefee? Token amounts are BigInts, [effectively bounded at 91 bits](https://github.com/filecoin-project/fvm/issues/56) given the current maximum supply in AttoFiL (unit of exchange).

File: rust/src/fvm/machine.rs [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    catch_panic_response(|| {
+        init_log();
+
+        info!("fil_create_fvm_machine: start");
+
+        let mut response = fil_CreateFvmMachineResponse::default();
+
+        let config = get_default_config();
+        let chain_epoch = chain_epoch as ChainEpoch;
+        let token_amount = TokenAmount::from_u64(token_amount);
```
Comment (raulk) [inline rust/src/fvm/machine.rs:80]: Can't be u64, needs to be u128 if safe for FFI, or a tuple of (hi, lo) u64s.

File: rust/src/fvm/machine.rs [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+                    Ok(x) => x,
+                    Err(err) => {
+                        response.status_code = FCPResponseStatus::FCPUnclassifiedError;
+                        response.error_msg = rust_str_to_c_str(format!("{:?}", err));
+                        return raw_ptr(response);
+                    }
+                };
+
+                response.status_code = FCPResponseStatus::FCPNoError;
+                // FIXME: Return relevant fields of ApplyRet
```
Comment (raulk) [inline rust/src/fvm/machine.rs:206]: Can return ApplyRet as (DAG-)CBOR encoded.

File: rust/src/fvm/machine.rs [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+            todo!("execute message")
+        }
+        None => {
+            todo!("invalid machine id")
+        }
+    }
+
+    info!("fil_fvm_machine_flush_message: end");
+    // })
+}
```
Comment (raulk) [inline rust/src/fvm/machine.rs:243]: I don't think we have a "finish message" operation. What we finish is the _machine_, which flushes all effectively linked blocks to the blockstore, and returns the new state root CID.

@Stebalien do we expect the node to construct the receipts root, or do we want to accumulate receipts in the machine, to then return the root?

File: rust/src/fvm/types.rs [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    pub to_len: libc::size_t,
+    pub sequence: u64,
+    pub value: u64,
+    pub method_num: u64,
+    pub params_bytes: *const u8,
+    pub params_len: libc::size_t,
+    pub gas_limit: u64,
+    pub gas_fee_cap: u64,
+    pub gas_premium: u64,
+}
```
Comment (raulk) [inline rust/src/fvm/types.rs:64]: We can use a DAG-CBOR-encoded message (which is how it's transmitted in the network anyway), and deserialize it in the Rust-side FFI handler, instead of modelling it again.

File: rust/src/fvm/machine.rs (hunk 1 above)
Comment (Stebalien) [reply rust/src/fvm/machine.rs:50]: No upper limit for now. We probably also need an exception for cron.

And yeah, we need to measure it. Then 10x at least.

Comment (Stebalien) [reply rust/src/fvm/machine.rs:50]: Note: I'd actually base the max memory on the total callstack. So we'll configure some max pages here, then set max_pages to `max_pages - used_pages` when instantiating a new instance.

File: rust/src/fvm/machine.rs (hunk 4 above)
Comment (Stebalien) [reply rust/src/fvm/machine.rs:80]: We usually just use bigints.

File: rust/src/fvm/machine.rs (hunk 6 above)
Comment (Stebalien) [reply rust/src/fvm/machine.rs:243]: I expect the node to deal with that, honestly. Basically:

1. The VM handles the state-tree _only_.
2. The Node handles everything related to the blockchain.

**Comment by cryptonemo** [comment]
Currently requires https://github.com/filecoin-project/ref-fvm/pull/277

File: rust/src/proofs/types.rs [hunk 8]
```diff
@@ -862,7 +862,7 @@ impl Clone for fil_SealCommitPhase2Response {
 code_and_message_impl!(fil_SealCommitPhase2Response);
 
 #[repr(C)]
-#[derive(Clone, DropStructMacro)]
+#[derive(Clone, Default, DropStructMacro)]
```
Comment (cryptonemo) [inline rust/src/proofs/types.rs:865]: Good catch, thanks

File: rust/src/fvm/types.rs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+code_and_message_impl!(fil_CreateFvmMachineResponse);
+
+#[repr(C)]
+#[derive(DropStructMacro)]
+pub struct fil_FvmMachineExecuteResponse {
+    pub error_msg: *const libc::c_char,
+    pub status_code: FCPResponseStatus,
+    pub exit_code: u64,
+    pub return_ptr: *const u8,
```
[the remark this answers is not in the document]
Comment (vmx) [reply rust/src/fvm/types.rs:48]: I think it's correct, it follows the usual pattern.

File: .circleci/config.yml [hunk 10]
```diff
@@ -7,7 +7,7 @@ orbs:
 executors:
   golang:
     docker:
-      - image: circleci/golang:1.13
+      - image: circleci/golang:1.16.4
```
Comment (mvdan) [inline .circleci/config.yml:10]: FYI this version is significantly out of date in terms of bugfix releases, the latest is 1.16.15. 1.16.x is also unsupported now that 1.18 is out, but I imagine you're already aware of that.

File: cgo/blockstore.go [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+	"unsafe"
+
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+)
+
+/*
+#include <stdint.h>
+typedef const uint8_t* buf_t;
```
Comment (dignifiedquire) [inline cgo/blockstore.go:13]: Why the use of `buf_t`? seems like this typedef should rather just live in go land if you want to use it

File: cgo/extern.go [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+	}
+}
+
+//export cgo_extern_verify_consensus_fault
+func cgo_extern_verify_consensus_fault(
+	handle C.uint64_t,
+	h1 C.buf_t, h1Len C.int32_t,
+	h2 C.buf_t, h2Len C.int32_t,
+	extra C.buf_t, extraLen C.int32_t,
+	minerId *C.uint64_t,
```
Comment (dignifiedquire) [inline cgo/extern.go:69]: would be more readable to prefix these out variables with sth like `out_`

File: cgo/registry.go [hunk 13]
```diff
@@ -0,0 +1,49 @@
+package cgo
+
+import (
+	"context"
+	"sync"
+)
+
+var (
+	mu       sync.RWMutex
```
Comment (dignifiedquire) [inline cgo/registry.go:9]: @Stebalien are you planning on pulling the atomic version into this?

File: fvm.go [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+	executor unsafe.Pointer
+}
+
+const (
+	applyExplicit = iota
+	applyImplicit
+)
+
+// CreateFVM
+func CreateFVM(fvmVersion uint64, externs cgo.Externs, epoch abi.ChainEpoch, baseFee abi.TokenAmount, baseCircSupply abi.TokenAmount, nv network.Version, stateBase cid.Cid) (*FVM, error) {
```
Comment (dignifiedquire) [inline fvm.go:38]: can we maybe use a config struct here, these are a lot of args

File: fvm.go [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+	baseFeeHi, baseFeeLo, err := splitBigInt(baseFee)
+	if err != nil {
+		return nil, xerrors.Errorf("invalid basefee: %w", err)
+	}
+	baseCircSupplyHi, baseCircSupplyLo, err := splitBigInt(baseCircSupply)
+	if err != nil {
+		return nil, xerrors.Errorf("invalid circ supply: %w", err)
+	}
+
+	exHandle := cgo.Register(context.TODO(), externs)
```
Comment (dignifiedquire) [inline fvm.go:48]: what's the plan regarding contexts here?

File: fvm.go [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+		f.executor = nil
+		generated.FilDropFvmMachine(executor)
+		cgo.Unregister(exHandle)
+	})
+
+	return fvm, nil
+}
+
+func (f *FVM) ApplyMessage(msgBytes []byte, chainLen uint) (*ApplyRet, error) {
+	defer runtime.KeepAlive(f)
```
Comment (dignifiedquire) [inline fvm.go:87]: why is this needed?

File: fvm.go [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+}
+
+// returns hi, lo
+func splitBigInt(i big.Int) (hi uint64, lo uint64, err error) {
+	if i.Sign() < 0 {
+		return 0, 0, xerrors.Errorf("negative number: %s", i)
+	}
+	words := i.Bits()
+	switch bits.UintSize {
+	case 32:
```
Comment (dignifiedquire) [inline fvm.go:166]: filecoin-ffi until now does not support 32bit systems, do we need to do so now?

File: rust/Cargo.toml [hunk 18]
```diff
@@ -38,22 +38,36 @@ memmap = "0.7"
...
 
 [features]
 default = ["opencl", "multicore-sdr" ]
 blst-portable = ["bls-signatures/blst-portable", "blstrs/portable"]
-opencl = ["filecoin-proofs-api/opencl", "bellperson/opencl", "storage-proofs-porep/opencl", "rust-gpu-tools/opencl"]
-cuda = ["filecoin-proofs-api/cuda", "bellperson/cuda", "storage-proofs-porep/cuda", "rust-gpu-tools/cuda"]
+opencl = ["filecoin-proofs-api/opencl", "bellperson/opencl", "storage-proofs-porep/opencl", "rust-gpu-tools/opencl", "fvm/opencl"]
+cuda = ["filecoin-proofs-api/cuda", "bellperson/cuda", "storage-proofs-porep/cuda", "rust-gpu-tools/cuda", "fvm/cuda"]
 multicore-sdr = ["storage-proofs-porep/multicore-sdr"]
+bindgen=[]
```
Comment (dignifiedquire) [inline rust/Cargo.toml:73]: what's this feature for?

File: rust/src/fvm/blockstore/cgo.rs [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+
+use anyhow::{anyhow, Result};
+use cid::Cid;
+use fvm_shared::blockstore::Blockstore;
+
+use super::super::cgo::*;
+
+const ERR_NO_STORE: i32 = -1;
+const ERR_NOT_FOUND: i32 = -2;
+const MAX_BUF_SIZE: usize = 4 << 20; // 4MiB
```
Comment (dignifiedquire) [inline rust/src/fvm/blockstore/cgo.rs:11]: can we get some comments as to what these constants exactly specify?

File: rust/src/fvm/blockstore/cgo.rs [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+                    e => Err(anyhow!("cgo blockstore 'put' failed with error code {}", e)),
+                }
+            }
+        }
+
+        let mut lengths = Vec::with_capacity(MAX_BLOCK_BATCH);
+        let mut buf = Vec::with_capacity(MAX_BUF_SIZE);
+        for (k, block) in blocks {
+            let block = block.as_ref();
+            if lengths.len() >= MAX_BLOCK_BATCH
```
Comment (dignifiedquire) [inline rust/src/fvm/blockstore/cgo.rs:105]: why is this needed?

File: rust/src/fvm/blockstore/fake.rs [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+    fn get(&self, k: &Cid) -> Result<Option<Vec<u8>>> {
+        match self.fake_blocks.borrow().get(k) {
+            Some(blk) => Ok(Some(blk.clone())),
+            None => self.base.get(k),
+        }
+    }
+
+    fn put_keyed(&self, k: &Cid, block: &[u8]) -> Result<()> {
+        // TODO: we won't need this hack once we merge
+        // https://github.com/filecoin-project/ref-fvm/pull/382.
```
Comment (dignifiedquire) [inline rust/src/fvm/blockstore/fake.rs:39]: this is merged, can we remove it?

File: rust/src/fvm/blockstore/overlay.rs [hunk 22]
```diff
@@ -0,0 +1,69 @@
+use std::collections::HashMap;
+
+use anyhow::Result;
+use cid::Cid;
+use fvm_shared::blockstore::Blockstore;
+
+pub struct OverlayBlockstore<BS> {
```
Comment (dignifiedquire) [inline rust/src/fvm/blockstore/overlay.rs:7]: please add some docs on what the purpose of this blockstore is

File: rust/src/fvm/externs.rs [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+                e => Err(anyhow!(
+                    "cgo extern 'verify_consensus_fault' failed with error code {}",
+                    e
+                )),
+            }
+        }
+    }
+}
+
+impl Externs for CgoExterns {}
```
Comment (dignifiedquire) [inline rust/src/fvm/externs.rs:128]: what's the purpose of this trait?

File: rust/src/fvm/machine.rs [hunk 24]
```diff
@@ -0,0 +1,276 @@
...
+    DefaultKernel<DefaultCallManager<DefaultMachine<OverlayBlockstore<CgoBlockstore>, CgoExterns>>>,
+>;
+
+lazy_static! {
+    static ref ENGINE: fvm::machine::Engine = fvm::machine::Engine::default();
+}
+
+/// Note: the incoming args as u64 and odd conversions to i32/i64
+/// for some types is due to the generated bindings not liking the
+/// 32bit types as incoming args
```
Comment (dignifiedquire) [inline rust/src/fvm/machine.rs:32]: This is very suspicious, and easy to get wrong. What were the issues you had here?

File: rust/src/fvm/machine.rs [hunk 25]
```diff
@@ -0,0 +1,276 @@
...
+        // Only do this if the return data is non-empty. The empty vec pointer is non-null and not
+        // valid in go.
+        if !apply_ret.msg_receipt.return_data.is_empty() {
+            let return_bytes = Vec::from(apply_ret.msg_receipt.return_data).into_boxed_slice();
+            response.return_ptr = return_bytes.as_ptr();
+            response.return_len = return_bytes.len();
+            Box::leak(return_bytes);
+        }
+
+        // TODO: Do something with the backtrace.
```
Comment (dignifiedquire) [inline rust/src/fvm/machine.rs:197]: is this expected to be done before being merged?

File: rust/src/fvm/machine.rs [hunk 26]
```diff
@@ -0,0 +1,276 @@
...
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn fil_drop_fvm_machine(executor: *mut libc::c_void) {
+    let _ = Box::from_raw(executor as *mut Mutex<CgoExecutor>);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn fil_fvm_machine_execute_message(
+    executor: *mut libc::c_void,
```
Comment (dignifiedquire) [inline rust/src/fvm/machine.rs:143]: why is this a void pointer, instead of a pointer to a concrete type?

File: cgo/blockstore.go (hunk 11 above)
Comment (Stebalien) [reply cgo/blockstore.go:13]: IIRC, I had issues with the "const" part. Although maybe we don't need that? This came from the "just work dammit" phase of development.

**Review by arajasek — APPROVED** [review]
Looked at this with fresh-ish eyes, and it seems good.

File: cgo/errors.go [hunk 27]
```diff
@@ -0,0 +1,8 @@
+package cgo
+
+const (
+	ErrInvalidHandle = -1 - iota
```
Comment (dignifiedquire) [inline cgo/errors.go:4]: does this  mean the next one is `0` or `-2`?

Comment (dignifiedquire) [reply cgo/errors.go:4]: I would suggest just defining them as their cgo versions, this way these are guranteed to match

File: cgo/registry.go (hunk 13 above)
Comment (dignifiedquire) [reply cgo/registry.go:9]: I meant this one that you wrote: https://github.com/filecoin-project/cgo-blockstore/blob/master/blockstore.go#L17

File: fvm.go [hunk 28]
```diff
@@ -0,0 +1,206 @@
...
+	baseFeeHi, baseFeeLo, err := splitBigInt(opts.BaseFee)
+	if err != nil {
+		return nil, xerrors.Errorf("invalid basefee: %w", err)
+	}
+	baseCircSupplyHi, baseCircSupplyLo, err := splitBigInt(opts.BaseCircSupply)
+	if err != nil {
+		return nil, xerrors.Errorf("invalid circ supply: %w", err)
+	}
+
+	exHandle := cgo.Register(context.TODO(), opts.Externs)
```
Comment (dignifiedquire) [inline fvm.go:58]: pass the context to `CreateFVM`?

File: bls.go [hunk 29]
```diff
@@ -1,8 +1,10 @@
-//+build cgo
+//go:build cgo
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply bls.go:1]: The go build flags format changed, so go now uses _both_ syntaxes (automatically).

File: fvm.go (hunk 28 above)
Comment (Stebalien) [reply fvm.go:58]: Let's reconsider this later. `NewVM` in lotus _does_ take a context, but it's a context for the construction of the VM.

File: cgo/errors.go (hunk 27 above)
Comment (Stebalien) [reply cgo/errors.go:4]: This is -1 as iota starts at 0.

File: rust/src/fvm/types.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Stebalien) [reply rust/src/fvm/types.rs:48]: This is dropped along with the rest of the response. The generated glue is:

```rust
impl Drop for fil_FvmMachineExecuteResponse {
    fn drop(&mut self) {
        unsafe {
            free_c_str(self.error_msg as *mut libc::c_char);
            drop(Vec::from_raw_parts(
                self.return_ptr as *mut u8,
                self.return_len,
                self.return_len,
            ));
        };
    }
}
```

File: cgo/registry.go (hunk 13 above)
Comment (Stebalien) [reply cgo/registry.go:9]: Oh, hell no. Not unless this turns out to be the bottleneck.

File: rust/src/fvm/types.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Stebalien) [reply rust/src/fvm/types.rs:48]: I'm not getting any segfaults, but likely won't wait for the test to limp to a halt.

File: bls.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply bls.go:1]: thanks

**Review by cryptonemo — APPROVED** [review]
